apis/placement/v1beta1/clusterresourceplacement_types.go

```diff
@@ -218,11 +218,51 @@ type PreferredClusterSelector struct {
 	Preference ClusterSelectorTerm `json:"preference"`
 }
 
+// MetricRange describes a required or preferred range for a metric.
+type MetricRange struct {
+	// Minimum is the minimum value of the range.
+	// If this value is not specified, the smallest observed value of the metric will
+	// be used.
+	// At least one of the minimum or maximum value must be specified.
+	// +optional
+	Minimum string `json:"minimum"`
+	// Maximum is the maximum value of the range.
+	// If this value is not specified, the largest observed value of the metric will be
+	// used.
+	// At least one of the minimum of maximum value must be specified.
+	// +optional
+	Maximum string `json:"maximum"`
+}
+
+// MetricRange is one metric requirement when picking clusters for resource placement.
+type MetricMatcher struct {
+	// Name is the name of the metric; it should be a Kubernetes label name.
+	// +required
+	Name string `json:"name"`
+	// Ranges are the required or preferred ranges for the metric.
+	// +required
+	Ranges []MetricRange `json:"ranges"`
+}
```

Comment:
How do we express that we want to find a cluster with a metrics key with the largest/smallest value using this?

Reply:
Yeah, that's actually a question I would like to ask before adding the APIs:

Earlier, it has been proposed that we put metric selectors as a part of the existing cluster selector terms, i.e. a cluster selector term = some label selectors + some metric selectors. This makes great sense as it reduces confusion and achieves the use case where user might want to filter out/prefer clusters with both specific labels and specific metrics with the right values. But it also brings out a side effect: the weight is an upper level field for a cluster selector term, i.e.,

```yaml
type PreferredClusterSelector struct {
	Weight int32 `json:"weight"`
	Preference ClusterSelectorTerm `json:"preference"`
}
```

Reply:
Basically there are two complications:

* User might want to express prefer maximum/minimum semantics using left/right open brackets, e.g., [10, max.) or (min, 30). This would normally require two weights (one for the min/max. value and one for the set limit), yet currently there's only one weight to specify. We could either add another weight field, which makes the API a bit weird if the user only uses label selectors; or we could assume that the weight for the set limit is always 0 ([10, max) -> cluster with 10 gets a score of 0, cluster with max gets the full weight), but this is also less than satisfactory as cluster with 10 receives the same score of 0 as clusters with a value of the range.
* We do allow multiple range expressions for a metric. The weight field being in the upper level means that all ranges have to use the same weight, with or without prefer min/max semantics.

Reply:
The most straightforward solution is just to add a interpolate from min/max switch, and assume that a) one end of the range always receives the weight of 0 with prefer min./max. semantics and b) all range expressions of a metric use the same weight -> would this be OK?